### ISSUE TITLE
Add support for Proxy Protocol (v2, binary) on incoming connections

### ISSUE THREAD

**Issue by frittentheke** [chriskohlhoff/asio#1091]
I first opened this issue with [Beast:](http://www.boost.org/libs/beast) at https://github.com/boostorg/beast/issues/2484,
but this kind of functionality actually lies in the handling of incoming TCP connections and that's done by asio in this case ...

> 
> I'd like to seek you to look into supporting proxy protocol v2 which allows proxy servers / TCP load balancers to forward information  about a client to the backend.
> 
> HAProxy does supports and actually originates this binary protocol specifying all sorts of fields to be piggybacked to the first TCP packet to the backend. This is instead of using HTTP headers the likes of `X-Forwarded-For`, `X-Forwarded-Proto`, which obviously only work when speaking HTTP and have other drawbacks.
> 
>  * https://www.haproxy.com/blog/use-the-proxy-protocol-to-preserve-a-clients-ip-address/
>  * https://github.com/haproxy/haproxy/blob/master/doc/proxy-protocol.txt (**includes sample code**)
> 
> But apart from HAProxy there are many more load balancers providing source info via proxy protocol v2 on outgoing connections to their backends. To not just name services from AWS, Azure, GCP which do support it for TCP connections via their load balancers or other networking services which terminate a TCP connection and would therefore "NAT" away source info.
> 
> The proxy protocol v2 can also carry info like client certificates in case of TLS offloading and last but lot least is extensible via custom TLVs (type, length, value). Implementations of the proxy protocol v2 sometimes internally replace data like source ip and port with the values received via proxy protocol v2 and treat the connection as it was directly established this way. So logging, rate limits, ACLs and other uses of this data then transparently uses the actual client info and are not reduced to only seeing the load balancer as client.
> 
> To give you a bit of back story ... I am a user of Ceph which uses Beast in their S3/Swift object storage service RGW (Rados Rateway),
> see:
>  * https://docs.ceph.com/en/quincy/radosgw/frontends/#beast
>  * https://github.com/ceph/ceph/blob/main/src/rgw/rgw_asio_frontend.cc
> 
> RGW is usually deployed in multiple instances and placed behind a load balancer, like HAProxy. And while it's certainly possible to have HAProxy (https://cbonte.github.io/haproxy-dconv/2.0/configuration.html#option%20forwardfor) or and other HTTP reverse proxy to send headers which Ceph RGW using beast then logs (`rgw_log_http_headers`) this does not work for load balancers in TCP.

**Comment by frittentheke** [comment]
> Why wouldn't this be it's own library?

@klemens-morgenstern  I don't have any expectations on how or where this would be packaged up or made into a library - I was already showing my confusion when opening the issue with Beast first. Please kindly assist in placing this feature request properly.

I myself started from the very end of the food chain (Ceph RGW -> Beast -> ASIO -> ...) and was just missing the data richness the proxy protocol v2 usually gives backend services which are behind load balancers.

**Comment by klemens-morgenstern** [comment]
Are you asking for an asio based implementation of haproxy? That would be a massive feature request that is unlikely to get done.

I don't understand why you wouldn't just use haproxy directly since it's on the tcp layer.

**Comment by frittentheke** [comment]
No, totally no. I believe there is a major misunderstanding in what I asked for.

I'd just like asio to be able to read the proxy protocol v2 info from incoming TCP connections. Connections that potentially originate from an instance of HAProxy or the many many more load balancers (including all the Cloud loadbalancer services) that support it to provide information about the actual client / source to a backend. 
And if the backend service (like Ceph RGW in my case) then uses ASIO internally it can access this info to make use of it in logs or to apply ACLs, rate-limiters or whatnot knowing what the actual source is.

Maybe this video helps explaining how the data is embedded in the TCP payload - https://www.youtube.com/watch?v=gT7ZK4hZAYU&t=735s.
